### PR TITLE
feat(baseline): Enable semgrep to perform baseline scans

### PR DESCRIPTION
**Pull request by brendongo**
This commit introduces the `--baseline-commit` flag. When given a commit hash (or valid git identifies like `HEAD^1`),
`semgrep` will only show findings that exist now but did not exist at said hash.

Semgrep accomplishes this by performing an initial scan,  checking out the older state, performing another scan, and
then removing findings from the former that appear in the latter. It considers findings the same if they have the same rule_id, path, and syntactic context (the lines that match the semgrep rule). Because of this we avoid refiring on findings when they simply have been moved around the file. However a limitation of this is that if a finding matches on a large section of lines (with a `...`, for example) some file changes will still show up as a `new` finding.

This behavior also interacts with inline ignores. Table below summarizes what we think it means for the 9 possible cases. `semgrep --baseline-commit` will only show findings that are either `New` or `Unignored`.

|                         |                                  |                                       |                                                      |
| ------------- | ------------------- | ---------------------- | ------------------------------- |
|                        | Finding in Baseline  | No Finding in Baseline | Finding in Baseline with nosem |
| Finding in Head  | Unchanged  | New  |  Unignored |
| No Finding in Head  | Fixed  | Unchanged | Fixed |
| Finding in Head with Nosem | Ignored | New and Ignored| Unchanged |

PR checklist:
- [ ] Documentation is up-to-date
- [ ] Changelog is up-to-date
- [ ] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

**Comment by aryx** [comment]
Haha, the license/cla got blocked on you :)

**Comment by brendongo** [comment]
Will need to open an issue to handle renames as baseline equivalent

File: semgrep/semgrep/core_output.py
Comment (brendongo) [inline semgrep/semgrep/core_output.py:357]: These were unused so got rid of them as part of refactor to make RuleMatch.lines usable with baseline scanning

File: semgrep/semgrep/core_runner.py
Comment (brendongo) [inline semgrep/semgrep/core_runner.py:415]: profiler was unused

File: semgrep/semgrep/git.py
Comment (brendongo) [inline semgrep/semgrep/git.py:1]: Mostly copies from https://github.com/returntocorp/semgrep-action/blob/5cb31fb22ea0fffbae63b5ad0986d42caa2b562c/src/semgrep_agent/targets.py with necessary changes to work in semgrep-cli

File: semgrep/semgrep/semgrep_main.py
Comment (brendongo) [inline semgrep/semgrep/semgrep_main.py:144]: Refactored into function so could be reused to scan head and baseline

**Comment by brendongo** [comment]
Still missing e2e tests but unit tests already included

File: semgrep/semgrep/git.py
Comment (emjin) [inline semgrep/semgrep/git.py]: I'm extremely confused about what this description means

File: semgrep/semgrep/semgrep_main.py
Comment (emjin) [inline semgrep/semgrep/semgrep_main.py]: Does this need to be in a try-catch? Or are errors here handled somewhere else? If so please update main docstring

**Review by emjin — APPROVED** [review]
Approved assuming we will add more tests

File: semgrep/semgrep/semgrep_main.py
Comment (brendongo) [reply semgrep/semgrep/semgrep_main.py]: Good call

**Comment by xmo-odoo** [comment]
Question about this PR: does it have some sort of cache which can be pre-populated with the baseline scans in order to avoid the extra validations?

e.g. let's say there's a bunch of PR against a base branch, by default `--baseline-commit` would have to perform a full scan of the base branch in all PRs in order to compute a baselined diagnostic report. But with some setup and caching the project could compute and store the base branch's report (probably useful anyway as a statistics / tracking device), then retrieve that from the per-PR jobs and compute things from that baseline directly.

**Pull request merged**
